dali/pipeline/executor/executor2/exec2.cc

```diff
@@ -0,0 +1,459 @@
+    SetupStreams();
+    SetupThreadPool();
+
+    last_iter_data_ = InitIterationData(-1);
+    if (last_iter_data_->checkpoint)
+      PopulateInitialCheckpoint(*last_iter_data_->checkpoint);
+
+    state_ = State::Built;
+    Start();
+  }
+
+  void Run() {
+    DeviceGuard dg(config_.device.value_or(CPU_ONLY_DEVICE_ID));
+    if (state_ != State::Running)
+      throw std::runtime_error("The executor is not initialized.");
+    InitIteration();
+    pending_outputs_.push(graph_.Launch(*exec_));
+  }
+
+  void Prefetch() {
+    DeviceGuard dg(config_.device.value_or(CPU_ONLY_DEVICE_ID));
+    for (int i = 0; i < prefetch_depth_; i++) {
+      Run();
+    }
+  }
+
+  Workspace PopOutputs() {
+    if (pending_outputs_.empty())
+      throw std::out_of_range("All pending outputs were already popped.");
+    DeviceGuard dg(config_.device.value_or(CPU_ONLY_DEVICE_ID));
+    auto fut = std::move(pending_outputs_.front());
+    pending_outputs_.pop();
+    auto &pipe_out = fut.Value<const PipelineOutput &>();
+    auto ws = pipe_out.workspace;
+    last_iter_data_ = ws.GetIterationData();
+    if (ws.has_event())
+      CUDA_CALL(cudaEventSynchronize(ws.event()));
+    ws.set_event(nullptr);
+    return ws;
+  }
+
+  void InitIteration() {
+    WorkspaceParams params{};
+    params.max_batch_size = config_.max_batch_size;
+    params.iter_data = InitIterationData(iter_index_++);
+    graph_.PrepareIteration(params);
+  }
+
+  int InputFeedCount(std::string_view) {
+    return prefetch_depth_;
+  }
+
+  OperatorBase *GetOperator(std::string_view input_name) const {
+    auto it = node_map_.find(input_name);
+    if (it == node_map_.end())
+      return nullptr;
+    return it->second->op.get();
+  }
+
+  const SharedIterData &LastIterData() const {
+    return last_iter_data_;
+  }
+
+  void Shutdown() {
+    DeviceGuard dg(config_.device.value_or(CPU_ONLY_DEVICE_ID));
+    if (state_ != State::Running)
+      return;
+    state_ = State::ShutdownRequested;
+    if (exec_)
+      exec_->Shutdown();
+    state_ = State::ShutDown;
+  }
+
+  void RestoreFromCheckpoint(const Checkpoint &cpt) {
+    DeviceGuard dg(config_.device.value_or(CPU_ONLY_DEVICE_ID));
+    int restored = 0;
+    for (auto &n : graph_.Nodes()) {
+      if (n.op) {
+        n.op->RestoreState(cpt.GetOpCheckpoint(n.instance_name));
+        restored++;
+      }
+    }
+    if (cpt.NumOp() > restored) {
+      throw std::runtime_error("The checkpoint data contains superfluous operator states.");
+    }
+    this->iter_index_ = cpt.GetIterationId();
+    if (!last_iter_data_)
+      last_iter_data_ = InitIterationData(-1);
+    last_iter_data_->checkpoint = std::make_shared<Checkpoint>(cpt);
+  }
+
+  void EnableCheckpointing(bool enabled) {
+    config_.checkpointing = enabled;
+  }
+
+  bool CheckpointingEnabled() const {
+    return config_.checkpointing;
+  }
+
+ private:
+  // Must be 1st member to be destroyed last.
+  std::optional<DeviceGuard> dtor_guard_;
```

Comment:
Is it possible to add some test/static_assert that will check this condition? I'm asking, because somebody may by accident add a member above (e.g. in unlabelled `private` section right after `class` keyword)

Also, how this would relate if somebody adds a static variable e.g in the constructor?

Reply:
I hope the comment will serve to address the former. I could add a base class, but there's nothing that could stop a careless contributor from adding another one before that.

As for the second - well, that would be extremely bad programming - and adding a static variable that depends on the thread-local state (current device) is asking for trouble to say the least.